### ISSUE TITLE
Lint suggestion: warn on `use` statements already imported via the prelude

### ISSUE THREAD

**Issue by burtonageo** [Manishearth/rust-clippy#1124]
A large number of traits, types and functions are imported into a rust crate implicitly through the [standard prelude](https://doc.rust-lang.org/1.9.0/std/prelude/). However, when you explicitly bring one of the items in the prelude into scope using the `use` statement, the rust compiler doesn't warn that the `use` is unnecessary.

 A clippy warning for this could help shave unnecessary lines. This should probably not run on crates with #![no_std]

**Comment by yaahc** [comment]
This came up in our libs team yesterday when discussing https://github.com/rust-lang/rust/pull/82786. Josh [summarized our conclusion](https://github.com/rust-lang/rust/pull/82786#issuecomment-819791302) but the gist is we want to rely on inherent traits but we expect that inherent traits will have a similar issue to items in the prelude where it's possible to have unused imports that don't produce warnings.

Our suggestion was to add an allow by default lint to warn about these unnecessary imports, and to have it work for both prelude items and inherent traits.

**Comment by edwloef** [comment]
Since the inherent traits RFC's FCP has been cancelled for some time with no further activity, is this relevant again? I'd like to make use of such a lint as well.

**Comment by bushrat011899** [comment]
This could interact negatively with conditional `#![no_std]`. For example:

```rust
#![cfg_attr(not(feature = "std"), no_std)]

extern crate alloc;

use alloc::boxed::Box;

pub struct Foo(pub Box<u8>);
```

Currently this works fine, but with a lint for unnecessary `use` statements `use alloc::boxed::Box;` will intermittently emit a warning based on whether `std` is enabled or not. The fix would be to require such crates to move to unconditional `no_std`:

```rust
#![no_std]

extern crate alloc;
#[cfg(feature = "std")]
extern crate std;

use alloc::boxed::Box;

pub struct Foo(pub Box<u8>);
```

_However_, based on [this comment](https://github.com/rust-lang/rust-clippy/pull/16971#discussion_r3208373385), forcing that style change would be undesirable. So the only option would be to conditionally import `Box`/allow the lint:

```rust
#![cfg_attr(not(feature = "std"), no_std)]

extern crate alloc;

#![cfg(not(feature = "std")] // or #[allow(clippy::already_imported_through_prelude)]
use alloc::boxed::Box;

pub struct Foo(pub Box<u8>);
```
